File: src/lib/rusto/src/base_expected.rs
use crate::base_check::BaseCheck;
use crate::types::IbddInAny;
use std::collections::HashMap;

/// BaseExpected extends BaseCheck for TDT pattern.
/// Validates each row in table-driven testing.
pub struct BaseExpected<I: IbddInAny> {
    inner: BaseCheck<I>,
    expected_value: Option<Box<dyn std::any::Any>>,
}

impl<I: IbddInAny> BaseExpected<I> {
    pub fn new(name: String, expected_cb: Box<dyn Fn(I::Iselection) -> I::Then>) -> Self {
        Self {
            inner: BaseCheck::new(name, expected_cb),
            expected_value: None,
        }
    }
    
    pub fn add_artifact(&mut self, path: String) {
        self.inner.add_artifact(path);
    }
    
    pub fn to_obj(&self) -> HashMap<String, Box<dyn std::any::Any>> {
        self.inner.to_obj()
    }
    
    // Set expected value for current row
    pub fn set_expected_value(&mut self, expected: Box<dyn std::any::Any>) {
        self.expected_value = Some(expected);
    }
    
    // Validate current row
    pub async fn validate_row(
        &mut self,
        store: I::Istore,
        test_resource: &dyn std::any::Any,
        filepath: &str,
        expected_value: Box<dyn std::any::Any>,
    ) -> Result<I::Then, String> {
        self.set_expected_value(expected_value);
        self.inner.test(store, test_resource, filepath).await
    }
    
    // Alias test to expect for TDT pattern
    pub async fn expect(
        &mut self,
        store: I::Istore,
        test_resource: &dyn std::any::Any,
        filepath: &str,
        expected_value: Box<dyn std::any::Any>,
    ) -> Result<I::Then, String> {
        self.validate_row(store, test_resource, filepath, expected_value).await
    }
}

File: src/lib/tiposkripto/src/verbs/tdt/BaseExpected.ts
import type { TestTypeParams_any } from "../../CoreTypes.js";
import type { ITestResourceConfiguration } from "../../types.js";

/**
 * BaseExpected for TDT pattern - independent implementation
 * Validates each row in table-driven testing.
 */
export abstract class BaseExpected<I extends TestTypeParams_any> {
  name: string;
  expectedCB: (val: I["iselection"]) => Promise<I["action"]>;
  expectedValue: any = null;
  error: Error | null = null;
  status: boolean | undefined;

  constructor(
    name: string,
    expectedCB: (val: I["iselection"]) => Promise<I["action"]>,
  ) {
    this.name = name;
    this.expectedCB = expectedCB;
  }

  // Set expected value for current row
  setExpectedValue(expected: any) {
    this.expectedValue = expected;
  }

  /**
   * Abstract method to validate a row
   */
  abstract validateRow(
    store: I["istore"],
    testResourceConfiguration: any,
    filepath: string,
    expectedValue: any,
    artifactory?: any,
  ): Promise<any>;

  async test(
    store: I["istore"],
    testResourceConfiguration: ITestResourceConfiguration,
    filepath: string,
    artifactory?: any,
  ): Promise<any> {
    try {
      const result = await this.expectedCB(store as any);
      this.status = true;
      return result;
    } catch (e: any) {
      this.status = false;
      this.error = e;
      throw e;
    }
  }

  toObj() {
    return {
      name: this.name,
      status: this.status,
      error: this.error ? `${this.error.name}: ${this.error.message}` : null,
      expectedValue: this.expectedValue,
    };
  }
}

export type IExpecteds<I extends TestTypeParams_any> = Record<string, BaseExpected<I>>;

// Export the BaseExpected class as named export
export { BaseExpected };
